drivers: Check errors on devices selected using dt macro

erwango opened the pull request:
Some Kconfig defined devices may be defined using dt_chosen_label
function. Since there is no way to ensure a device enabled in dts
is also defined in Kconfig, it may happen that instance is not
actually defined.
In this case device_get_binding might return 0, leading to undefined
behavior in the function that calls it.
When not already done, systematically check return of function
device_get_binding on devices defined through dt_chosen_label macro.
Trigger ASSERT when required and return error when possible.

Fixes #20068

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

joerchan commented on subsys/bluetooth/host/monitor.c:
```diff
@@ -328,6 +328,8 @@ static int bt_monitor_init(struct device *d)
 
 	monitor_dev = device_get_binding(CONFIG_BT_MONITOR_ON_DEV_NAME);
 
+	__ASSERT_NO_MSG(monitor_dev);
```
If asserts are not enabled we would give monitor_dev to uart_irq_rx/tx_disable below. Maybe we should do the handling that you did in shell_uart.c in all places that you added the assert?

erwango replied:
This was my initial intention, but in functions invoked in `SYS_INIT` macro (like current `bt_monitor_init`), returned values won't be checked AFAIU.

joerchan replied:
That is strange, seems wierd to have return codes then. But the return would stop you from doing something harmful at least, it could be a null pointer write which would lead to MPU fault or other types of memory errors. And maybe somebody will check the return value in the future :)

erwango replied:
> But the return would stop you from doing something harmful at least, it could be a null pointer write which would lead to MPU fault or other types of memory errors. 

I'd rather not do this as this would likely be easily caught by various code analysis scripts, so this would add support follow-up to this PR.

> And maybe somebody will check the return value in the future :)

Or... somebody will require these functions to be void :-)

joerchan replied:
well I'm just skeptical since the ASSERT might not be enabled. And then the check would simply be skipped an the code continues, so we would end up in undefined behavior anyway.

erwango replied:
I agree. ASSERT are disabled by default, but in case one started to debug this kind of things, enabling them would be one of the first step (was mine at least).
Anyway the real solution to this is solving Kconfig/dt desynchro and @galak is working on it.

nashif commented on subsys/shell/shell_uart.c:
```diff
@@ -274,6 +274,11 @@ static int enable_shell_uart(struct device *arg)
 		CONFIG_LOG_MAX_LEVEL : CONFIG_SHELL_BACKEND_SERIAL_LOG_LEVEL;
 
+	if (dev == NULL) {
+		__ASSERT_NO_MSG(dev);
```
having asserts and error checking in the same spot is not ideal IMO, you either assert on the whole thing or do error handling and propagation.

erwango replied:
Agreed. Fixed, thanks for review!

erwango committed:
drivers: susbsys: Check errors on devices selected using dt macro

Some Kconfig defined devices may be defined using dt_chosen_label
function. Since there is no way to ensure a device enabled in dts
is also defined in Kconfig, it may happen that instance is not
actually defined.
In this case device_get_binding might return 0, leading to undefined
behavior in the function that calls it.
When not already done, systematically check return of function
device_get_binding on devices defined through dt_chosen_label macro.
Trigger ASSERT when required and return error when possible.

Fixes #20068

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@nashif updated

erwango commented:
@nashif, can you review again ?

joerchan commented:
@erwango Misspelled subsys? (susbsys). Anyway I don't think subsys should be in the title, so you could remove it.

erwango commented:
Even if the initial issue which is dts/Kconfig device instance desynchro that was at the initiation of this issue cannot happen anymore due to recent changes, I think the check is sane and worth keeping.
If this does not go in v2.3.0, I'll close the point so we don't keep it eternally.
